apps/emqx_resource/src/emqx_resource_manager.erl

```diff
@@ -0,0 +1,408 @@
+
+%% @doc Lookup the group and data of a resource
+-spec lookup(instance_id()) -> {ok, resource_group(), resource_data()} | {error, not_found}.
+lookup(InstId) ->
+    safe_call(InstId, lookup).
+
+%% @doc Lookup the group and data of a resource
+-spec ets_lookup(instance_id()) -> {ok, resource_group(), resource_data()} | {error, not_found}.
+ets_lookup(InstId) ->
+    case ets:lookup(?ETS_TABLE, InstId) of
+        [{_id, Group, Data}] ->
+            {ok, Group, data_record_to_external_map(Data)};
+        [] ->
+            {error, not_found}
+    end.
+
+%% @doc Reset the metrics for the specified resource
+-spec reset_metrics(instance_id()) -> ok.
+reset_metrics(InstId) ->
+    emqx_plugin_libs_metrics:reset_metrics(resource_metrics, InstId).
+
+%% @doc Returns the data for all resorces
+-spec list_all() -> [resource_data()].
+list_all() ->
+    try
+        [
+            data_record_to_external_map(Data)
+         || {_Id, _Group, Data} <- ets:tab2list(?ETS_TABLE)
+        ]
+    catch
+        error:badarg -> []
+    end.
+
+%% @doc Returns a list of ids for all the resources in a group
+-spec list_group(resource_group()) -> [instance_id()].
+list_group(Group) ->
+    List = ets:match(?ETS_TABLE, {'$1', Group, '_'}),
+    lists:flatten(List).
+
+-spec health_check(instance_id()) -> ok | {error, Reason :: term()}.
+health_check(InstId) ->
+    safe_call(InstId, health_check).
+
+%% Server start/stop callbacks
+
+%% @doc Function called from the supervisor to actually start the server
+start_link(InstId, Group, ResourceType, Config, Opts) ->
+    Data = #data{
+        id = InstId,
+        group = Group,
+        mod = ResourceType,
+        config = Config,
+        opts = Opts,
+        status = undefined,
+        state = undefined,
+        error = undefined
+    },
+    gen_statem:start_link({local, proc_name(InstId)}, ?MODULE, Data, []).
+
+init(Data) ->
+    {ok, connecting, Data}.
+
+terminate(_Reason, _State, Data) ->
+    ets:delete(?ETS_TABLE, Data#data.id),
+    ok.
+
+%% Behavior callback
+
+callback_mode() -> [handle_event_function, state_enter].
+
+%% Common event Function
+
+%% Called when the resource needs to be stopped
+handle_event({call, From}, set_resource_status_connecting, _State, Data) ->
+    {keep_state, Data#data{status = connecting}, [{reply, From, ok}]};
+handle_event({call, From}, stop, _State, #data{status = disconnected} = _Data) ->
+    {keep_state_and_data, [{reply, From, ok}]};
+handle_event({call, From}, stop, _State, Data) ->
+    Result = do_stop(Data),
+    {next_state, disconnected, Data, [{reply, From, Result}]};
+handle_event({call, From}, {remove, ClearMetrics}, _State, #data{id = InstId} = Data) ->
+    do_stop(Data),
+    ets:delete(?ETS_TABLE, InstId),
+    case ClearMetrics of
+        true -> ok = emqx_plugin_libs_metrics:clear_metrics(resource_metrics, InstId);
+        false -> ok
+    end,
+    {stop_and_reply, normal, [{reply, From, ok}]};
```

Comment:
Nit: one way to improve readability of state machines is to move away complicated business logic from `handle_event` function to a separate internal function. E.g. `handle_remove(From, ClearMetrics, Data)`